Package requirements and guidelines

eduardo-rodrigues opened the pull request:
@henryiii, you may remember the discussions we had weeks ago to try and dress a list of requirements and guidelines for the Scikit-HEP packages. I finally tried to turn the discussions into a PR!

The idea would be to have this as a first submenu under https://scikit-hep.org/developer. Some fine-tuning will be needed. Let's try and converge. In particular, some links are likely to be missing, even links to your other pages under https://scikit-hep.org/developer.

Thanks.

eduardo-rodrigues commented on pages/developers/guidelines.md:
```diff
@@ -0,0 +1,47 @@
+### Code formatting.
+
+We suggest as a guideline to use [Black][], which is rather popular these days.
+See [/style][] for details.
```
This is most probably the wrong way to refer to the same-level page https://scikit-hep.org/developer/style. I wrote a guess, I confess. Please amend if needed ...

henryiii replied:
I think there's a filter you should use, but I have to first see which version of Jekyll we use - this was simplified in 4.0.

eduardo-rodrigues replied:
OK, sounds good.

eduardo-rodrigues replied:
@henryiii, any idea, then? You probably use such a feature in the IRIS-HEP pages?

henryiii replied:
I’ll check in ~5 mins

henryiii replied:
Looks like GH pages is still stuck on 3.x. I'll update these with the 3.x syntax.

eduardo-rodrigues commented:
I forgot to say, @henryiii: an alternative to this new page would be to have the guidelines under the page https://scikit-hep.org/developer. I have no strong feeling.

henryiii commented on pages/developers/guidelines.md:
```diff
@@ -0,0 +1,47 @@
+---
+layout: page
+title: Package requirements and guidelines
+permalink: /guidelines
```
This should be nested, too, in the dev folder, like the other dev pages

eduardo-rodrigues replied:
You mean it should be `permalink: /developer/guidelines`? About to commit the correction.

henryiii commented on pages/developers/guidelines.md:
```diff
@@ -0,0 +1,47 @@
+# Package requirements and guidelines
+
+The below provides requirements and guidelines for Scikit-HEP packages,
+which we hope help providing high-quality packages with some level of homogeneity across them.
```
I would add a line like: New packages should adhere to these minimum requirements. Further optional suggestions are in the following pages." or something like that.

eduardo-rodrigues replied:
Good point. Added.

henryiii commented on pages/developers/guidelines.md:
```diff
@@ -0,0 +1,47 @@
+As a  guideline, we try to be as open as possible with licensing,
+following what most of the other packages use in the wide Python scientific community;
+Scikit-HEP packages typically use the BSD-3 license.
+It is acknowledged that exceptions are needed and "sticky licenses" such as GPL may be the only viable way.
```
```suggestion
It is acknowledged that exceptions are needed and "sticky licenses" such as GPL may be the only viable option.
```

henryiii commented on pages/developers/guidelines.md:
```diff
@@ -0,0 +1,47 @@
+
+A reasonable test suite is a requirement, and it is desirable to achieve high coverage with clever tests - naturally.
+We recommend to use [pytest][] as a testing framework
+and GitHub Actions (GHA) for how to run the tests in continuous integration (CI).
```
Let's add a cross link here. Once I verify the correct format.

henryiii commented:
I've fixed the links, and I've updated the page a bit, feel free to check the changes.

eduardo-rodrigues commented:
Thanks @henryiii for the updates and improvements. I made a slight mess when updating on my side but believe I fixed it all with my last commit.

I made tiny changes to what you updated/changed. The main one is because I really feel that we should say "requirements and guidelines", not just requirements, which can sound harsh. Also, I removed "new" because we also strive to follow these guidelines as much as possible for the packages already in the org.

This is ready to be merged. Thanks again.
